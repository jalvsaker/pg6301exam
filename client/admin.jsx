import { useNavigate } from "react-router-dom";
import { useEffect, useState } from "react";
import { AddFood } from "./addFood";

export function Admin({ user }) {
  const navigate = useNavigate();

  if (!user.isAdmin) {
    navigate("/");
  }

  const [foods, setFoods] = useState([]);
  const [loading, setLoading] = useState(true);
  const [reloader, setReloader] = useState(false);

  useEffect(() => {
    (async () => {
      const res = await fetch("/api/foods");
      if (res.ok) {
        setFoods(await res.json());
        setLoading(false);
      }
    })();
  }, [reloader]);

  function reload() {
    setReloader((r) => !r);
  }

  if (loading) {
    return <div>Loading...</div>;
  }

  return (
    <>
      <h1>Admin panel</h1>

      <AddFood reload={reload} />

      {foods.map((food) => (
        <div key={food._id}>{food.name}</div>
      ))}
    </>
  );
}
