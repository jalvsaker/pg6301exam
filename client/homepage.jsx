export function Homepage() {
  return <h1>Hello world</h1>;
}
