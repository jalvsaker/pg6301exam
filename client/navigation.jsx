import { Link, NavLink } from "react-router-dom";
import React from "react";

export function Navigation({ user }) {
  return (
    <nav>
      <NavLink to={"/"}>Home</NavLink>
      {!user.username && <NavLink to={"/login"}>Log In</NavLink>}
      {user.username && (
        <span>
          {user.username} <Link to={"/logout"}>Log out</Link>
        </span>
      )}
    </nav>
  );
}
