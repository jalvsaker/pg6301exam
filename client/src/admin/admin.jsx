import { useNavigate } from "react-router-dom";
import { useEffect, useState } from "react";
import { AddFood } from "./addFood";
import { ChangeFood } from "./changeFood";

export function Admin({ user }) {
  const navigate = useNavigate();

  if (!user.isAdmin) {
    navigate("/");
  }

  const [foods, setFoods] = useState([]);
  const [loading, setLoading] = useState(true);
  const [reloader, setReloader] = useState(false);

  useEffect(() => {
    (async () => {
      const res = await fetch("/api/foods");
      if (res.ok) {
        setFoods(await res.json());
        setLoading(false);
      }
    })();
  }, [reloader]);

  function reload() {
    setReloader((r) => !r);
  }

  if (loading) {
    return <div>Loading...</div>;
  }

  return (
    <>
      <h1>Admin panel</h1>

      <h3>Add new foods:</h3>
      <AddFood reload={reload} />

      <h3>Change previous foods:</h3>

      {foods.map((food) => (
        <ChangeFood key={food._id} food={food} reload={reload}/>
      ))}
    </>
  );
}
