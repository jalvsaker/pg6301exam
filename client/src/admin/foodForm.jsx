import React from "react";
import { useState } from "react";

export function FoodForm({ onSubmit, food, buttonText, resetAfter }) {
  const [name, setName] = useState(food?.name || "");
  const [price, setPrice] = useState(food?.price || "");
  const [description, setDescription] = useState(food?.description || "");

  async function handleSubmit(e) {
    e.preventDefault();

    const body = { name, price, description };

    await onSubmit(body);

    if (resetAfter) {
      setName("");
      setPrice("");
    }
  }

  return (
    <form onSubmit={handleSubmit}>
      <div>
        <label>
          Name:
          <input
            type={"text"}
            value={name}
            onChange={(e) => setName(e.target.value)}
          />
        </label>
      </div>
      <div>
        <label>
          Price:
          <input
            type={"number"}
            value={price}
            onChange={(e) => setPrice(e.target.value)}
          />
        </label>
      </div>
      <div>
        <label>
          Description:
          <textarea
            cols="30"
            rows="4"
            value={description}
            onChange={(e) => setDescription(e.target.value)}
          ></textarea>
        </label>
      </div>
      <button>{buttonText}</button>
    </form>
  );
}
