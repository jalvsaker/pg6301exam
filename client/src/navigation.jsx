import { Link, NavLink } from "react-router-dom";
import React from "react";

export function Navigation({ user }) {
  return (
    <nav>
      <NavLink to={"/"}>Home</NavLink>
      {!user.username && <NavLink to={"/login"}>Log In</NavLink>}
      {!user.username && <NavLink to={"/register"}>Register new user</NavLink>}
      {user.username && (
        <>
          <span>Welcome {user.username}</span>
          <Link to={"/logout"}>Log out</Link>
        </>
      )}
      {user.isAdmin && <NavLink to={"/admin"}>Admin panel</NavLink>}
    </nav>
  );
}
