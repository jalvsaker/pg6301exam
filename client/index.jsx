import React from "react";
import { createRoot } from "react-dom/client";
import {
  BrowserRouter,
  Navigate,
  NavLink,
  Route,
  Routes,
} from "react-router-dom";
import { Homepage } from "./homepage";

function App() {
  return (
    <BrowserRouter>
      <nav>
        <NavLink to={"/"}>Home</NavLink>
        <NavLink to={"/test"}>Test</NavLink>
      </nav>
      <Routes>
        <Route path={"/"} element={<Homepage />} />
        <Route path={"/test"} element={<h2>test</h2>} />
        <Route path={"/*"} element={<Navigate to={"/"} />} />
      </Routes>
    </BrowserRouter>
  );
}

const root = createRoot(document.getElementById("root"));
root.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
);
