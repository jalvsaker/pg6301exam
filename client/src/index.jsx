import React, { useEffect, useState } from "react";
import { createRoot } from "react-dom/client";
import { BrowserRouter, Navigate, Route, Routes } from "react-router-dom";
import { Homepage } from "./homepage";
import { Login } from "./login/login";
import { Navigation } from "./navigation";
import { Logout } from "./login/logout";
import { Register } from "./login/register";
import { Menu } from "./menu/menu";
import { Admin } from "./admin/admin";
import "./styles.css";
import { Chat } from "./chat/chat";
import { ShowOrders } from "./admin/showOrders";

function App() {
  const [user, setUser] = useState({});
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    (async () => {
      const res = await fetch("/api/login");
      if (res.ok) {
        setUser((await res.json()) || {});
      }
      setLoading(false);
    })();
  }, []);

  if (loading) {
    return "Loading...";
  }

  return (
    <BrowserRouter>
      <Navigation user={user} />
      <Routes>
        <Route path={"/"} element={<Homepage />} />
        <Route path={"/menu"} element={<Menu user={user} />} />
        <Route path={"/register"} element={<Register setUser={setUser} />} />
        <Route path={"/login"} element={<Login setUser={setUser} />} />
        <Route path={"/logout"} element={<Logout setUser={setUser} />} />
        <Route path={"/admin"} element={<Admin user={user} />} />
        <Route path={"/admin/orders"} element={<ShowOrders user={user} />} />
        <Route path={"/chat"} element={<Chat user={user} />} />
        <Route path={"/*"} element={<Navigate to={"/"} />} />
      </Routes>
    </BrowserRouter>
  );
}

const root = createRoot(document.getElementById("root"));
root.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
);
