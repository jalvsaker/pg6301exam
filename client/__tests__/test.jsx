import React from "react";
import { createRoot } from "react-dom/client";
import { Navigation } from "../src/navigation";
import { act, Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router-dom";
import { Homepage } from "../src/homepage";
import { FoodForm } from "../src/admin/foodForm";

describe("tests", () => {
  it("should show nav menu", async () => {
    const element = document.createElement("div");
    const root = createRoot(element);

    await act(() => {
      root.render(
        <MemoryRouter>
          <Navigation user={{ username: "name", isAdmin: true }} />
        </MemoryRouter>
      );
    });

    expect(element.innerHTML).toMatchSnapshot();
  });

  it("should show homepage", async () => {
    const element = document.createElement("div");
    const root = createRoot(element);

    await act(() => {
      root.render(
        <MemoryRouter>
          <Homepage />
        </MemoryRouter>
      );
    });

    expect(element.innerHTML).toMatchSnapshot();
  });

  it("should submit food form", async () => {
    const element = document.createElement("div");
    const root = createRoot(element);

    const mock = jest.fn();

    await act(() => {
      root.render(<FoodForm onSubmit={mock}/>);
    });

    await act(()=>{
      Simulate.change(element.getElementsByTagName("input")[0], {
        target: { value: "foodname" },
      });
      Simulate.change(element.getElementsByTagName("input")[1], {
        target: { value: "12" },
      });

    })

    act(()=>{
      Simulate.submit(element.querySelector("form"));
    })

    expect(element.innerHTML).toMatchSnapshot();
    expect(mock).toBeCalledWith({name: "foodname", price: "12"})
  });
});
