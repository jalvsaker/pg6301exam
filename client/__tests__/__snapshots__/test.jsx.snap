// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests should show Admin panel 1`] = `"<h1>Admin panel</h1><a href="/admin/orders">Show orders</a><hr><h3>Add new foods:</h3><form><div><label>Name:<input type="text" value=""></label></div><div><label>Price:<input type="number" value=""></label></div><button>Add</button></form><h3>Change previous foods:</h3>"`;

exports[`tests should show addFood 1`] = `"<form><div><label>Name:<input type="text" value=""></label></div><div><label>Price:<input type="number" value=""></label></div><button>Add</button></form>"`;

exports[`tests should show changeFood 1`] = `"<form><div><label>Name:<input type="text" value=""></label></div><div><label>Price:<input type="number" value=""></label></div><button>Update</button></form><button>Delete</button><hr>"`;

exports[`tests should show homepage 1`] = `"<h1>Bob's Catering</h1><p>Welcome to Bob's catering. Mediocre food to mediocre prices.</p><a href="/menu">See the menu</a>"`;

exports[`tests should show nav menu 1`] = `"<nav><a aria-current="page" class="active" href="/">Home</a><span>Welcome name</span><a href="/chat">Chat</a><a href="/logout">Log out</a><a class="" href="/admin">Admin panel</a></nav>"`;

exports[`tests should submit chat 1`] = `"<h1>Chat</h1><hr><div><strong>user 1</strong> - Hello<hr></div><div><strong>user 2</strong> - Hey<hr></div><form><input type="text" value=""><button>Send</button></form>"`;

exports[`tests should submit food form 1`] = `"<form><div><label>Name:<input type="text" value="foodname"></label></div><div><label>Price:<input type="number" value="12"></label></div><button></button></form>"`;
